velox/common/caching/AsyncDataCache.cpp

```diff
@@ -230,45 +223,25 @@ bool FusedLoad::loadOrFuture(folly::SemiFuture<bool>* wait) {
     state_ = LoadState::kLoading;
   }
   // Outside of 'mutex_'.
+  std::exception_ptr error;
   try {
-    // If wait is not set this counts as prefetch.
-    loadData(!wait);
-    for (auto& pin : pins_) {
-      pin.entry()->setValid();
-    }
-    {
-      std::lock_guard<std::mutex> l(mutex_);
-      pins_.clear();
+    auto pins = loadData(!wait);
+    for (auto& pin : pins) {
+      auto entry = pin.checkedEntry();
+      VELOX_CHECK(entry->key().fileNum.hasValue());
+      VELOX_CHECK(entry->isExclusive());
+      entry->setExclusiveToShared();
     }
     setEndState(LoadState::kLoaded);
-    return true;
-  } catch (const std::exception& e) {
-    cancel();
-    std::rethrow_exception(std::current_exception());
-  }
-}
-
-void FusedLoad::cancel() {
-  std::unique_ptr<folly::SharedPromise<bool>> promise;
-  {
-    std::lock_guard<std::mutex> l(mutex_);
-    if (state_ == LoadState::kLoading || state_ == LoadState::kCancelled) {
-      // Already cancelled or another thread is loading. If loading, we must let
-      // the load finish.
-      return;
-    }
-    state_ = LoadState::kCancelled;
-    promise = std::move(promise_);
-    for (auto& pin : pins_) {
-      if (!pin.empty()) {
-        pin.entry()->setValid(false);
-      }
-    }
-    pins_.clear();
+  } catch (std::exception& e) {
+    error = std::current_exception();
   }
-  if (promise) {
-    promise->setValue(true);
+  if (error) {
+    // Cleanup outside of catch to avoid killing the process if more exceptions.
```

Comment:
This is an unusual pattern. Why not wrap setEndState call in a separate try-catch?